File: python/1600. Throne Inheritance.py
class TreeNode(object):

    def __init__(self, name):
        self.name = name
        self.child = []
        self.dead = False

    def insert(self, parentName, childName):
        def insert(root):
            if not root:
                return
            if root.name==parentName:
                node = TreeNode(childName)
                root.child.append(node)
                return node
            for ch in root.child:
                insert(ch)
        return insert(self)


    def remove(self, name):
        def remove(root):
            if not root:
                return
            if root.name==name:
                root.dead = True
                return
            for ch in root.child:
                remove(ch)
        remove(self)

    def traverse(self):
        ans = []
        def traverse(root):
            if not root:
                return
            if not root.dead:
                ans.append(root)
            for ch in root.child:
                traverse(ch)
        traverse(self)
        return ans
    
    def present(self):
        return ' -> '.join(map(lambda node: node.name, self.traverse()))

class ThroneInheritance(object):

    def __init__(self, kingName):
        """
        :type kingName: str
        """
        self.root = TreeNode(kingName)
        self.mem = {kingName: self.root}


    def birth(self, parentName, childName):
        """
        :type parentName: str
        :type childName: str
        :rtype: None
        """
        node = self.mem[parentName].insert(parentName, childName)
        self.mem[childName] = node

    def death(self, name):
        """
        :type name: str
        :rtype: None
        """
        self.mem[name].dead = True

    def getInheritanceOrder(self):
        """
        :rtype: List[str]
        """
        return map(lambda root: root.name, self.root.traverse())



# Your ThroneInheritance object will be instantiated and called as such:
# obj = ThroneInheritance(kingName)
# obj.birth(parentName,childName)
# obj.death(name)
# param_3 = obj.getInheritanceOrder()


if __name__ == '__main__':
    root = TreeNode('King')
    root.insert('King', 'Alice')
    print root.present()
    # print root.name, root.child[0].name

File: go/src/leetcode/1600. Throne Inheritance/1600. Throne Inheritance.go
package main

type node struct {
	name string
	child []*node
}

type ThroneInheritance struct {
	root *node
	death map[string]int
	name2node map[string]*node
}


func Constructor(kingName string) ThroneInheritance {
	root := &node{name: kingName, child: []*node{}}
	return ThroneInheritance{
		root: root,
		death: make(map[string]int),
		name2node: map[string]*node{kingName: root},
	}	
}


func (this *ThroneInheritance) Birth(parentName string, childName string)  {
	p := this.name2node[parentName]
	newNode := &node{name: childName, child: []*node{}}
	this.name2node[childName] = newNode
	p.child = append(p.child, newNode)
}


func (this *ThroneInheritance) Death(name string)  {
	this.death[name]++
}


func (this *ThroneInheritance) GetInheritanceOrder() []string {
	order := []string{}
	this.successor(this.root, &order)
	return order
}

func (this *ThroneInheritance) successor(root *node, order *[]string) {
	if root == nil {
		return
	}
	if _, ok := this.death[root.name]; !ok {
		*order = append(*order, root.name)
	}
	for _, ch := range root.child {
		this.successor(ch, order)
	}
}


/**
 * Your ThroneInheritance object will be instantiated and called as such:
 * obj := Constructor(kingName);
 * obj.Birth(parentName,childName);
 * obj.Death(name);
 * param_3 := obj.GetInheritanceOrder();
 */
